mypy/server/aststripnew.py

```diff
@@ -85,13 +85,16 @@ def prepare_implicit_var_patches(self, node: ClassDef) -> None:
         """Produce callbacks that re-add attributes defined on self."""
         for name, sym in node.info.names.items():
             if isinstance(sym.node, Var) and sym.implicit:
+                explicit_self_type = sym.node.explicit_self_type
+
                 def patch() -> None:
                     existing = node.info.get(name)
+                    current_existing = node.info.names.get(name)
                     # This needs to mimic the logic in SemanticAnalyzer.analyze_member_lvalue()
                     # regarding the existing variable in class body or in a superclass.
-                    # TODO: copy the logic better (always create a Var if there is explicit type)
-                    if (not existing or
-                            isinstance(existing.node, Var) and existing.node.is_abstract_var):
+                    if (existing is None or
+                            isinstance(existing.node, Var) and existing.node.is_abstract_var or
```

Comment:
This check is not very clear. Maybe also check that `current_existing` is `None` here, i.e. something like this:

```
if (existing is None or
    (current_existing is None and ((isinstance(existing.node, Var) and existing.node.is_abstract_var) or
                                    explicit_self_type))):
```

Style nit: I prefer to use parens to make precedence explicit when combining `and` / `or` in a single expression.

Reply:
I didn't check because it is impossible to have `current_exiting` if `existing` is `None`.